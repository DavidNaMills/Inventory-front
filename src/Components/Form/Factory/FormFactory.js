import React from 'react';
import classes from './formFactory.module.scss';
import {useLocale} from '../../../Context/LocaleContext';

const FormFactory = (props) => {
    const locale = useLocale();
    const { config, displayFields, onChange } = props;
    let element = null;
    if (displayFields && displayFields[config.id]) {
        
        switch (config.elementType.toLowerCase()) {
            case 'input':

                element = <input
                    key={config.id}
                    id={config.id}
                    {...config.elementConfig}
                    placeholder= {locale[config.elementConfig.placeholder]}
                    label={locale[config.elementConfig.placeholder]}
                    onChange={e => onChange(e, config.id)}
                    value={config.value}
                    className={classes.input}
                />
                break;

            case 'select':
                element = < select
                    id={config.id}
                    onChange={e => onChange(e, config.id)}
                    className={classes.select}
                >
                    {config.elementConfig.options.map(x => {
                        return (
                            <option
                                key={x._id}
                                value={x._id}
                            >
                                {x.displayValue}
                            </option>
                        )
                    }
                    )}
                </select>
                break;

            case 'multiple':
                element = < select
                    id={config.id}
                    multiple
                    onChange={e => onChange(e, config.id)}
                    className={classes.select}
                >
                    {config.elementConfig.options.map(x => {
                        return (
                            <option
                                key={x._id}
                                value={x._id}
                            >
                                {x.displayValue}
                            </option>
                        )
                    }
                    )}
                </select>
                break;

            case 'image':
                element = <input
                    type = 'file'
                    onChange={(e) => onChange(e, config.id)}
                    placeholder={locale[config.elementConfig.placeholder]}
                    className={classes.input}
            />
                break;

            default:
                element = <p>{`${config.elementType} : ${config.elementConfig.placeholder}`}</p>
                break;
        }
    }   // end of if

    return (
        <React.Fragment>
            {
                element
                    ? <div className={classes.formFactory}>
                        <label htmlFor={config.id} className={classes.formFactory__label}>{locale[config.elementConfig.placeholder]}</label>
                        {element}
                        {
                            (config.errorMsg && config.errorMsg.length > 0 ) &&
                                <div className={classes.formFactory__errorMsg}>
                                    {config.errorMsg.map((e, c)=><p key={c}>{e}</p>)}
                                </div>
                        }
                    </div>
                    : null
            }
        </React.Fragment>
    )
}

export default FormFactory;