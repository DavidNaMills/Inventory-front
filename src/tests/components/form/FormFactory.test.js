import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import '../../testHelpers/context';

import FormFactory from '../../../Components/Form/Factory/FormFactory';

import { formFactoryTestData } from '../form/formConfigTestData';

// elementConfig: {
//     type: 'text',
//     placeholder: 'phone'
// },

const mockOnChange = jest.fn();

afterEach(() => {
    jest.clearAllMocks();
})

const displayFields = {
    phone: 'phone',
    numberTest: 'numberTest',
    password: 'password',
    image: 'image',
    role: 'role',
    baseId: 'baseId',
    fail: 'fail',
    withErrors: 'withErrors'
};

const baseProps = {
    config: formFactoryTestData.phone,
    displayFields,
    onChange: mockOnChange
}

describe('<FormFactory/> test suite', () => {
    it('matches snapshot', () => {
        const comp = renderer.create(<FormFactory {...baseProps} />);
        const app = comp.toJSON();
        expect(app).toMatchSnapshot();
    });

    it('exists', () => {
        const wrapper = shallow(<FormFactory {...baseProps} />);
        expect(wrapper.exists).toBeTruthy();
    });

    describe('form elements tests', () => {
        [
            {
                msg: 'renders an <input/> element and reacts to onChange',
                config: formFactoryTestData.phone,
                type: 'input',
                id: 'phone',
                calledWith: 'dsa123'
            },
            {
                msg: 'renders an <input/> element with number property and reacts to onChange',
                config: formFactoryTestData.numberTest,
                type: 'input',
                id: 'numberTest',
                calledWith: '123'
            },
            {
                msg: 'renders an <input/> element with password property and reacts to onChange',
                config: formFactoryTestData.password,
                type: 'input',
                id: 'password',
                calledWith: 'pwe45d'
            },

        ].forEach(x => {

            it(x.msg, () => {
                const newProps = { ...baseProps, config: x.config };
                const wrapper = shallow(<FormFactory {...newProps} />);
                expect(wrapper.find('input').length).toBe(1);
                wrapper.find('input').simulate('change', x.calledWith)
                expect(mockOnChange).toHaveBeenCalledTimes(1);
                expect(mockOnChange).toHaveBeenCalledWith(x.calledWith, x.id);
            });
        });

        it('renders a <select/> element and reacts to onChange', () => {
            const newProps = { ...baseProps, config: formFactoryTestData.role };
            const wrapper = shallow(<FormFactory {...newProps} />);
            expect(wrapper.find('select').length).toBe(1);
            wrapper.find('select').simulate('change', 'admin')
            expect(mockOnChange).toHaveBeenCalledTimes(1);
            expect(mockOnChange).toHaveBeenCalledWith('admin', 'role');
        });

        it('renders a <select/> element with multiple property and reacts to onChange', () => {
            const newProps = { ...baseProps, config: formFactoryTestData.baseId };
            const wrapper = shallow(<FormFactory {...newProps} />);
            expect(wrapper.find('select').length).toBe(1);
            wrapper.find('select').simulate('change', ['admin', 'tests', 'super tests'])
            expect(mockOnChange).toHaveBeenCalledTimes(1);
            expect(mockOnChange).toHaveBeenCalledWith(['admin', 'tests', 'super tests'], 'baseId');
            expect(wrapper.find('select').props()).toHaveProperty('multiple', true);
        });

        it('renders a <file/> element and reacts to onChange', () => {
            const newProps = { ...baseProps, config: formFactoryTestData.image };
            const wrapper = shallow(<FormFactory {...newProps} />);
            expect(wrapper.find('input').length).toBe(1);
            wrapper.find('input').simulate('change', ['test file'])
            expect(mockOnChange).toHaveBeenCalledTimes(1);
            expect(mockOnChange).toHaveBeenCalledWith(['test file'], 'image');
            expect(wrapper.find('input').props()).toHaveProperty('type', 'file');
        });

        it('renders a <p> tag if element is not supported', () => {
            const newProps = { ...baseProps, config: { id: 'fail', elementType: 'fail', elementConfig: { placeholder: 'test' } } };
            const wrapper = shallow(<FormFactory {...newProps} />);
            expect(wrapper.find('p').length).toBe(1);
        });
    });

    describe.only('display tests', () => {
        it('renders a <div> with class="formFactory", a label with class="formFactory__label", and the element', () => { 
            const wrapper = shallow(<FormFactory {...baseProps} />);
            expect(wrapper.find('div').props()).toHaveProperty('className', 'formFactory');
            expect(wrapper.find('label').props()).toHaveProperty('className', 'formFactory__label');
        });
        
        it('renders a <div> with class="formFactory__errorMsg" and a <p> tag for each error', () => {
            const newProps = { ...baseProps, config: formFactoryTestData.withErrors };
            const wrapper = shallow(<FormFactory {...newProps} />);
            
        });
    });
});