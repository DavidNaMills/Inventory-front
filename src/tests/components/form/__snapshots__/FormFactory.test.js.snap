// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormFactory/> test suite matches snapshot 1`] = `null`;
